### ISSUE TITLE
Missing classes with Scala incremental compilation

### ISSUE THREAD

**Issue by fteychene** [gradle/gradle#13392]
With incremental compilation on Scala project you can have a success build with missing classes in the destination dir in build folder. The main issue is that it goes to cache and break projects dependencies on it (tests or other projects) even with a clean.

### Expected Behavior
Build classes should not missing classes if compilation is a success with incremental compilation

### Current Behavior
In some cases (see steps to reproduce) we have a compilation success with a build dir that miss some classes

### Context
We have a mono repo of scala projects and developers issued compilation failing for test for classes not found or dependency on another project that compile. In bot cases classes exists and compileTask was succeeding or loaded from cache.
A lot of these issues where after a branch switch or a merge/rebase in feature branch.

### Steps to Reproduce 
[scala-zinc-cache.tar.gz](https://github.com/gradle/gradle/files/4758833/scala-zinc-cache.tar.gz)
there is a `reproduce.sh` script to have some change that lead to the missing classes

To reproduce :
 - `build` a project
 - Edit a source file to make it not compile
 - `compileScala`
 - Revert the modification and edit another file with a valid modification
 - `compileScala`
 - `build/classes/main` Does not contains the classes for the first edited file

### Your Environment

Reproduce on other projects with gradle 6.1 / 6.4 / 6.5
Build scan URL:
  - `build` -> https://gradle.com/s/ihiwuotevegl6
 - `compileScala` with voluntary compilation error - >  https://gradle.com/s/o6yzclizv57i2
 - `compileScala` success but with missing classes -> https://gradle.com/s/mrh3lqmiojeso

**Comment by msogrin** [comment]
I came here to post the same issue as OP, with pretty much the same reproduction steps, but using 6.3.
Sometimes, I could even run gradle clean and gradle assemble and get empty build/classes directory with hundreds of classes missing (but resources present).

And attempt to disable incremental compilation was also unsuccessful with very similar symptoms. But for that, there's an item #13224. 

Could these two items be related?  Could this item be resolved by the fix to #13224?

**Comment by ivan-klass** [comment]
Thank you @fteychene for finding easy reproduction steps! This problem affects us quite often, as I've also mentioned in comments to 13224

https://github.com/gradle/gradle/issues/13224#issuecomment-634190044

**Comment by ljacomet** [comment]
The fix for #13224 will not fix this at all. will try to get this issue fixed for 6.6

**Comment by ivan-klass** [comment]
> The fix for #13224 will not fix this at all. will try to get this issue fixed for 6.6

Thank you! If you find any sort of workaround to avoid the issue or fix the state after it's corrupted, please also share here :)

**Comment by msogrin** [comment]
I think the workaround from #13224 should also work here, but without a check for non-incremental mode.
(kotlin DSL)
```
tasks.withType<ScalaCompile>().configureEach {
    doLast {
        scalaCompileOptions.incrementalOptions.analysisFile.get().asFile.delete()
    }
}
```

**Comment by ljacomet** [comment]
Indeed, though that will kill all incrementality, so you probably want to do that only when you have an issue.

**Comment by ljacomet** [comment]
Note that if anyone in the community feels like looking into this issue, it would be very welcome. I can provide guidance on understanding how Gradle interacts with Zinc.

**Comment by SheliakLyr** [comment]
Great reproduction steps @fteychene !

I think this problem is caused by using the default "ClassfileManagerType". When not configured, DeleteImmediatelyManagerType is used, which on compile error leaves destination directory empty.

In order to keep previous compilation results in destinationDir (needed for keeping in sync with AnalysisFile), we need to use TransactionalManagerType. On compilation error, it reverts changes made in destinationDir.
```scala
IncOptions.of().withClassfileManagerType(
   Optional.of(
      TransactionalManagerType.of(
         new File("/tmp/classes.backup"), // use better file!
         new SbtLoggerAdapter())))
```    

[ZincScalaCompiler.java:127](https://github.com/gradle/gradle/blob/7d05c3b4228d2e60dbecc38bacecf4421ff7d80c/subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java) needs to be changed + some logic to define where backup file is located.

Mixing incremental compilation with distributed build cache scares me.. I suspect there might be other problems, like: 
- what happens to analysis store when compilation outputs are loaded from the cache? 
- what happens when analysis store cannot be updated due to I/O error?
- what happens when TransactionalManagerType fails to restore the destinationDir? (well.. something similar to what we see in this issue :) )

Maybe outputs should be stored in build cache only on full recompile to ensure that no "broken" cache entries are produced? Not ideal, but I prefer it to random compilation/runtime "ClassNotFounds". Another idea: verify that destinationDir is correct for given analysis results (which, as far I can see, has no hashes/timestamps for compilation results).

**Comment by msogrin** [comment]
@SheliakLyr thank you for the analysis, it looks very promising.

Here's an example how SBT does it https://github.com/sbt/sbt/blob/5ca27512ca2ab4651309e0495039e20fd095df19/main/src/main/scala/sbt/Defaults.scala#L716-L724
```scala
        .withClassfileManagerType(
          Option(
            TransactionalManagerType
              .of( // https://github.com/sbt/sbt/issues/1673
                crossTarget.value / s"${prefix(configuration.value.name)}classes.bak",
                sbt.util.Logger.Null
              ): ClassFileManagerType
          ).toOptional
        )
```
So no special handling for any kind of abnormal situations. It only ensures that parallel compilation  of multiple targets uses different files names.

**Comment by ijuma** [comment]
@SheliakLyr @msogrin any of you interested in submitting a PR with the fix?

**Comment by SheliakLyr** [comment]
I'm interested, just didn't have time.  I will have "production-grade" fix this week, maybe tommorow.

**Comment by ljacomet** [comment]
@SheliakLyr Thanks for offering your help there, it will be much appreciated!

**Comment by ljacomet** [comment]
The fix from @SheliakLyr has been merged and will be released with Gradle 6.6
